flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxImpl.java

```diff
@@ -0,0 +1,233 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks.mailbox;
+
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Implementation of {@link Mailbox} inspired by {@link java.util.concurrent.ArrayBlockingQueue} and tailored towards
+ * our use case with multiple writers, single reader and volatile reads instead of lock & read on {@link #count}.
+ */
+public class MailboxImpl implements Mailbox {
+
+	/**
+	 * The enqueued letters.
+	 */
+	@GuardedBy("lock")
+	private final Runnable[] ringBuffer;
+
+	/**
+	 * Lock for all concurrent ops.
+	 */
+	private final ReentrantLock lock;
+
+	/**
+	 * Condition that is triggered when the buffer is no longer empty.
+	 */
+	@GuardedBy("lock")
+	private final Condition notEmpty;
+
+	/**
+	 * Condition that is triggered when the buffer is no longer full.
+	 */
+	@GuardedBy("lock")
+	private final Condition notFull;
+
+	/**
+	 * Index of the ring buffer head.
+	 */
+	@GuardedBy("lock")
+	private int headIndex;
+
+	/**
+	 * Index of the ring buffer tail.
+	 */
+	@GuardedBy("lock")
+	private int tailIndex;
+
+	/**
+	 * Number of letters in the mailbox.
+	 */
+	@GuardedBy("lock")
+	private volatile int count;
+
+	/**
+	 * A mask to wrap around the indexes of the ring buffer. We use this to avoid ifs or modulo ops.
+	 */
+	private final int moduloMask;
+
+	public MailboxImpl() {
+		this(6); // 2^6 = 64
+	}
+
+	public MailboxImpl(int capacityPow2) {
+		final int capacity = 1 << capacityPow2;
+		Preconditions.checkState(capacity > 0);
+		this.moduloMask = capacity - 1;
+		this.ringBuffer = new Runnable[capacity];
+		this.lock = new ReentrantLock();
+		this.notEmpty = lock.newCondition();
+		this.notFull = lock.newCondition();
+	}
+
+	@Override
+	public boolean hasMail() {
+		return count > 0;
+	}
+
+	@Override
+	public Runnable tryTakeMail() {
+		final ReentrantLock lock = this.lock;
+		lock.lock();
+		try {
+			return isEmpty() ? null : takeInternal();
+		} finally {
+			lock.unlock();
+		}
+	}
+
+	@Nonnull
+	@Override
+	public Runnable takeMail() throws InterruptedException {
+		final ReentrantLock lock = this.lock;
+		lock.lockInterruptibly();
+		try {
+			while (isEmpty()) {
+				notEmpty.await();
+			}
+			return takeInternal();
+		} finally {
+			lock.unlock();
+		}
+	}
+
+	@Override
+	public void waitUntilHasMail() throws InterruptedException {
+		final ReentrantLock lock = this.lock;
+		lock.lockInterruptibly();
+		try {
+			while (isEmpty()) {
+				notEmpty.await();
+			}
+		} finally {
+			lock.unlock();
+		}
+	}
+
+	//------------------------------------------------------------------------------------------------------------------
+
+	@Override
+	public boolean tryPutMail(@Nonnull Runnable letter) {
+		final ReentrantLock lock = this.lock;
+		lock.lock();
+		try {
+			if (isFull()) {
```

Comment:
A little doubt raise here that: as i know the mail box is enqueued by the timer or checkpoint action right? if this `enqueue` action should check the capacity of the queue, does this means that the timer trigger of checkpoint trigger may be put failed and have to retry with blocking the main thread?

Reply:
`tryPutMail` is currently a dead unused method ([checkout this discussion](https://github.com/apache/flink/pull/8442#discussion_r284575775)). For checkpointing we will be using blocking variant of this method, which will have exactly the same semantic as before - one thing that we will need to double check is old behaviour vs current in case if we enqueue multiple checkpoint triggers.

Reply:
@Aitozi thanks for thanking this through, here is how I concluded that this is not problematic:
- If you check the code, checkpoints and timers are even currently not invoked from the "main thread", but via executors.
- Right now both can also block on their attempt to acquire the checkpointing lock, which is very similar in effect to blocking on the mailbox.
- We have non-blocking ways of emitting letters to the mailbox, if the need arises.
Does this resolve your concerns?

Reply:
Yes, my focus is on that do we need to block the enqueue operation if the queue is full, As before there is no check for the `backoff` (like the capacity here) of the `to do list` like timer and checkpoint. Although the default capacity of mailbox(64) is enough I think,  Because only the pending checkpoint and the first timer poll from timerservice should be enqueued to this queue IMO, but once this queue is full, do we need to block the main thread?

Reply:
> @Aitozi thanks for thanking this through, here is how I concluded that this is not problematic:
> 
> * If you check the code, checkpoints and timers are even currently not invoked from the "main thread", but via executors.
> * Right now both can also block on their attempt to acquire the checkpointing lock, which is very similar in effect to blocking on the mailbox.
> * We have non-blocking ways of emitting letters to the mailbox, if the need arises.
> 
> Does this resolve your concerns?

sorry i do not refresh to see your reply just now, I will check the code again.

Reply:
If you make the comparison, the current behavior of blocking directly on the checkpointing lock is essentially the same having a queue of size 1. I was thinking through this for a bit and there should no downside from this change for responsiveness, but if you think there is I would like to hear about it.